torch/fx/experimental/Partitioner.py

```diff
@@ -0,0 +1,151 @@
+        return self.node_list[-1]
+
+    def __str__(self) -> str:
+        line: str = 'partition id: ' + str(self.partition_id) + '\n'
+        line += 'all nodes in this partition: '
+        for node_index, node_name, node_op in self.node_list:
+            line += '(node_' + str(node_index) + ':' + node_name +  ',' + node_op + ') '
+        line += '\n'
+        line += 'parent partitions:'
+        for parent in self.parents:
+            line += 'partition_' + str(parent.partitition_id) + ' '
+        line += '\n'
+        line += 'children partitions:'
+        for child in self.children:
+            line +=  str(child.partition_id) + ' '
+        line += '\n'
+        line += 'input nodes: '
+        input_nodes = self.get_input_nodes()
+        for node_index, node_name, node_op in input_nodes:
+            line += '(node_' + str(node_index) + ':' + node_name +  ',' + node_op + ') '
+        line += '\n'
+        output_node = self.get_output_node()
+        line += 'output node: ' + '(node_' + str(output_node[0]) + ':' + output_node[1] + ',' + output_node[2] + ')\n'
+        return line
+
+class Partitioner:
+    def __init__(self) -> None:
+        self.partitions: List[Partition] = []
+
+    """
+    After a partitioner is created,
+    it first check if multiple types of devices (backends) are involved
+    So far, we assume that there is only one backend and one device with unlimited memory
+    """
+
+    def partition_graph(self, fx_module: GraphModule, devices: List[str]) -> None:
+        graph: Graph = fx_module.graph
+        """check if devices are different backend types"""
+        multiple_backend_types: bool = self.check_multiple_backend_types(devices)
+        """one node could be partitioned into multiple partitions
+           mapping {node_index: [partition_ids]}
+        """
+        node_to_partition_mapping: Dict[int, List[int]] = {}
+        """only one type of devices"""
+        if not multiple_backend_types:
+            node_to_partition_mapping = self.load_balanced_partition(graph, devices)
+        else:
+            """todo: multiple_type_of_devices"""
+            pass
+        self.do_partition(fx_module, node_to_partition_mapping)
+        return
+
+    def load_balanced_partition(self, graph: Graph, devices: List[str]) -> Dict[int, List[int]]:
+        """same type of devices"""
+        node_to_partition_mapping: Dict[int, List[int]] = {}
+        """only one device, so far, assume memory is unlimited. Otherwise, memory size should be checked"""
+        if len(devices) == 1:
+            node_to_partition_mapping = self.find_single_partition(graph, devices[0])
+        else:
+            """todo: multiple devices of same type"""
+            pass
+        return node_to_partition_mapping
+
+    def find_single_partition(self, graph: Graph, device: str) -> Dict[Node, List[int]]:
+        """only one partition (one graph on one device)"""
+        partition_0: Partition = self.create_partition(0)
```

Comment:
In general you should add type information to your function arguments but you don't need to add type information for every local variable.